partiql-lang/src/main/kotlin/org/partiql/lang/eval/PartiqlAstExtensions.kt

```diff
@@ -65,3 +65,114 @@ internal fun PartiqlAst.Expr.getStartingSourceLocationMeta(): SourceLocationMeta
+
+        // When a variable reference (a PartiqlAst.Expr.Id) is encountered, add it to the accumulator.
+        override fun walkExprId(node: PartiqlAst.Expr.Id, accumulator: Set<String>): Set<String> {
+            return accumulator + node.name.text
+        }
+
+        // The invariant of each walkExprXxx(node: Xxx, accumulator) call
+        // is to add all *free* variables of `node` expression to `accumulator`.
+        // The default implementations work for all expressions except SELECT.
+
+        // Processing the SELECT involves taking care that variable occurrences bound by FROM, LET, and GROUP BY clauses
+        // do not end up in the result, if they are in the scope of those bindings.
+        override fun walkExprSelect(node: PartiqlAst.Expr.Select, accumulator: Set<String>): Set<String> {
+            var current = emptySet<String>()
+
+            // SELECT clauses are processed in the reverse of evaluation order, since that's how variable scoping extends.
+
+            // Note: in `SELECT e AS x ...`, free variables of `e` get into the result,
+            // but `x` *does not*, due to `x` being a 'symbol', not an 'id' in the AST.
+            // (For reference, 'x' *should not* count as either free or bound because it is not a variable,
+            // it is a spec to create an attribute in a struct.)
+            current += walkProjection(node.project, emptySet())
+
+            node.limit?.let { current += walkExpr(it, emptySet()) }
+            node.offset?.let { current += walkExpr(it, emptySet()) }
+
+            node.order?.let { current += walkOrderBy(it, emptySet()) }
+
+            node.having?.let { current += walkExpr(it, emptySet()) }
+
+            node.group?.let {
+                val freeInGroup = walkGroupBy(it, emptySet())
+                val boundByGroup =
+                    (it.groupAsAlias?.let { setOf(it.text) } ?: emptySet()) +
+                        it.keyList.keys.flatMap { it.asAlias?.let { setOf(it.text) } ?: emptySet() }
+                current = (current - boundByGroup) + freeInGroup
+            }
+
+            node.where?.let { current += walkExpr(it, emptySet()) }
+
+            node.fromLet?.let {
+                current = addFreeOfLets(it.letBindings, current)
+            }
+
+            current = addFreeOfFrom(node.from, current)
+
+            return accumulator + current
+        }
+
+        // Accounts for situations when an earlier binder is referenced in a later FROM item, such as t in 2nd item in
+        //   FROM Tbl as t, t as x
+        // Such a reference should not be counted as free.
+        private fun addFreeOfFrom(n: PartiqlAst.FromSource, accum: Set<String>): Set<String> =
+            when (n) {
+                is PartiqlAst.FromSource.Join -> {
+                    val acc = addFreeOfFrom(n.right, accum)
+                    addFreeOfFrom(n.left, acc)
+                }
```

Comment:
Should we search the JOIN condition for free variables?

Reply:
Yes, this was an oversight; good catch!